### PR TITLE
feat (#new); cbridge; added celer cbridge subgraph

### PR DESCRIPTION
**Pull request by tnkrxyz**
This PR adds the cbridge subgraph using the bridge schema and sdk.

## Tasks/remaining issues
- [x] Percentage fee (revenue) varying by chain and token, and the information is only available via their State Guardian Network API (https://cbridge-docs.celer.network/developer/api-reference/gateway-estimateamt) and not available on-chain. I asked on celeb network discord server, there is no no-chain information available if/when these fees change.
    - The current implementation tracks fee withdrawal through pool-based bridge, but it is impossible to distinguish between protocol-side revenue (fee attributed to SGN) or supply-side revenue (fee attributed to liquidity providers)
- [x] cbridge pool-based bridge allows users to refund failed transfers (there are > 300 of refunds so far: https://dune.com/queries/1861535). How best to handle this with bridge sdk? 
    - refunds are handled with a transferIn that is reverse of transferOut.
- [x] certain events/calls miss necessary information, for example, no sender information is available for transferIn in pool-based bridge; no receiver or dstChainID info for PeggedTokenBridge burn event.
    - senders are assumed to have the same address as the receiver in those cases 
- [x] Since the pool entity in the bridge schema only support one inputToken, it is necesary to create one (virtual) pool for each token in the pool-based bridge; which is fine (with a minor issue of having many pools), but it creates a problem of not being able to allocate liquidity farming rewards to these pools. 
    - A pool is created for each pool and token combination 
- [ ] ~~rename to celer-bridge after review approval~~

### boba (new)
  - **Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-boba**
  - **Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-boba&tab=protocol**
### aurora (fixed addresses)
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-aurora
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-aurora&tab=protocol
### celo (fixed addresses)
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-celo
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-celo&tab=protocol
### harmony (fixed addresses)
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-harmony
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-harmony&tab=protocol
### moonbeam (fixed addresses)
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-moonbeam
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonbeam&tab=protocol

### ethereum
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-ethereum
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-ethereum&tab=protocol
### arbitrum
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-arbitrum
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-arbitrum&tab=protocol
### avalanche
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-avalanche
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-avalanche&tab=protocol
### bsc
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-bsc
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-bsc&tab=protocol
### celo
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-celo
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-celo&tab=protocol
### fantom
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-fantom
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-fantom&tab=protocol
### gnosis
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-gnosis
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-gnosis&tab=protocol
### harmony 
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-harmony
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-harmony&tab=protocol
### moonbeam
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-moonbeam
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonbeam&tab=protocol
### moonriver
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-moonriver
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonriver&tab=protocol
### optimism
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-optimism
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-optimism&tab=protocol
### polygon
  - Test deployment: https://thegraph.com/hosted-service/subgraph/tnkrxyz/cbridge-polygon
  - Validation dashboard: https://subgraphs.messari.io/subgraph?endpoint=https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-polygon&tab=protocol

### CODE REVIEW

**Review by pietydespises — APPROVED** [review]

**Comment by pietydespises** [comment]
@jaimehgb please review when you get a chance

**Comment by pietydespises** [comment]
@jaimehgb can you review these code changes if they are still relevant?

**Comment by pietydespises** [comment]
Hey @tnkrxyz , I'm not able to view the validation dashboards for any of the deployments except for Ethereum and Fantom.

Getting the following error: " DEPLOYMENT ERROR - https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonbeam does not have any "protocol" entities. Essential data that determines validation can not be pulled without this entity. "

**Comment by tnkrxyz** [comment]
> Hey @tnkrxyz , I'm not able to view the validation dashboards for any of the deployments except for Ethereum and Fantom.
> 
> Getting the following error: " DEPLOYMENT ERROR - https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonbeam does not have any "protocol" entities. Essential data that determines validation can not be pulled without this entity. "

Looks like all of them got stuck at different % after deployed 3 weeks ago because of problem of the hosted service. I will try to re-deploy to see I can get it syncing again. Currently experience the same issue with another subgraph (bastion) too; looks like wide-spread problem not isolated to cbridge or a single chain.

**Comment by tnkrxyz** [comment]
> > Hey @tnkrxyz , I'm not able to view the validation dashboards for any of the deployments except for Ethereum and Fantom.
> > Getting the following error: " DEPLOYMENT ERROR - https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-moonbeam does not have any "protocol" entities. Essential data that determines validation can not be pulled without this entity. "
> 
> Looks like all of them got stuck at different % after deployed 3 weeks ago because of problem of the hosted service. I will try to re-deploy to see I can get it syncing again. Currently experience the same issue with another subgraph (bastion) too; looks like wide-spread problem not isolated to cbridge or a single chain.

OK, the problem is that hosted service doesn't support callHandlers for chains other than a few: mainnet, fantom, etc. Since the cbridge subgraph heavily utilizes callHandlers, it won't work on these chains. I will discuss with @jaimehgb what to do in this situation.

**Comment by pietydespises** [comment]
Ok @tnkrxyz thanks for the update, lmk when it needs to be checked

**Comment by tnkrxyz** [comment]
> Ok @tnkrxyz thanks for the update, lmk when it needs to be checked

@pietydespises I came up with a potential solution with @jaimehgb that I will test and implement this week. Will keep you posted.

**Review by pietydespises — DISMISSED** [review]

File: subgraphs/cbridge/package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "license": "MIT",
+  "scripts": {
+    "format": "npx prettier --write ."
+  },
+  "dependencies": {
+    "@graphprotocol/graph-cli": "0.37.0",
```
Comment (dmelotik) [inline subgraphs/cbridge/package.json:7]: use latest npm versions

File: subgraphs/cbridge/src/mappings.ts [hunk 2]
```diff
@@ -0,0 +1,1187 @@
...
+    event.params.dstChainId,
+    event.params.sender,
+    event.params.receiver,
+    event.params.message,
+    event.params.fee,
+    customEvent
+  );
+}
+
+export function handleMessageWithTransfer(event: MessageWithTransfer): void {
```
Comment (dmelotik) [inline subgraphs/cbridge/src/mappings.ts:705]: Does this do anything else? ie, do we need to use `_handleTransferOut()`?

File: subgraphs/cbridge/src/mappings.ts [hunk 3]
```diff
@@ -0,0 +1,1187 @@
...
+      ]);
+      return [BigInt.fromI32(v), BigInt.fromI32(endIdx)];
+    }
+  }
+
+  throw new Error("Invalid varint stream"); // i=10, invalid varint stream
+}
+
+// decode bytes
+function decodeBytes(buf: Bytes, idx: i32): Bytes[] | null {
```
Comment (dmelotik) [inline subgraphs/cbridge/src/mappings.ts:1127]: Can we make this generalizable and add it to the _reference_ subgraph in a separate PR for future use?

Comment (dmelotik) [reply subgraphs/cbridge/src/mappings.ts:1127]: with a method header on what it does

**Review by dmelotik — COMMENTED** [review]
Great job @tnkrxyz ! 🤌 Code is solid, nice commenting

- Is the call handler solution done and working as expected?
- Can you fix the deployment.json merge conflict?
- I know this PR is from a while ago, but we do have a new bridge schema now, it is not too much more to support. Do you think it is worthwhile to add that?

**Comment by tnkrxyz** [comment]
> Great job @tnkrxyz ! 🤌 Code is solid, nice commenting

thx!
 
>     * Is the call handler solution done and working as expected?

Yes, we have a solution for this, by using the mustache template tag {{^supportCallHandlers}} and setting defaults when we have no choice but the event.
 
>     * Can you fix the deployment.json merge conflict?

done.
 
>     * I know this PR is from a while ago, but we do have a new bridge schema now, it is not too much more to support. Do you think it is worthwhile to add that?

good point! upgraded.

File: subgraphs/cbridge/src/mappings.ts (hunk 2 above)
Comment (tnkrxyz) [reply subgraphs/cbridge/src/mappings.ts:705]: good question. My reading of their contract is that users will need to do a separate `sendTokenTransfer` call when sendMessageWithTransfer, which will emit a transfer out event (which event depending on the tyep of bridge used): https://github.com/celer-network/sgn-v2-contracts/blob/9ce8ffe13389415a53e2c38838da1b99689d40f0/contracts/message/libraries/MessageSenderLib.sol#L66-L98.

I added a comment for this.

File: subgraphs/cbridge/src/mappings.ts (hunk 3 above)
Comment (tnkrxyz) [reply subgraphs/cbridge/src/mappings.ts:1127]: Actually the byte decoding method is tied to cbridge only - I am not aware any other protocol packing data this way: https://github.com/celer-network/sgn-v2-contracts/blob/aa569f848165840bd4eec8134f753e105e36ae38/contracts/libraries/PbPool.sol#L20-L45

Comment (dmelotik) [reply subgraphs/cbridge/src/mappings.ts:1127]: ah, understood. Resolving

**Review by dmelotik — APPROVED** [review]

File: subgraphs/cbridge/package.json (hunk 1 above)
Comment (tnkrxyz) [reply subgraphs/cbridge/package.json:7]: removed. I don't think we need package.json at the subgraphs/cbridge folder.

**Comment by tnkrxyz** [comment]
@dmelotik , I just found a few mistakes I made in configurations.json (wrong contract addresses & network names) for celo, harmony, and moonbeam. I also added boba. can you pls help take a look? thx!

**Comment by dmelotik** [comment]
@tnkrxyz the changes LGTM

- There are some changes to arrakis and uniswap v3 arbitrum in the deployment.json though
- Do we need to re-QA check the deployments that got fixed?
- Also, in Boba (WAGMI v3 Option), the `mintSupply` is negative. Shouldn't it be only >= 0. I don't think it makes sense to say that the minted supply on the chain is < 0

**Comment by tnkrxyz** [comment]
> * There are some changes to arrakis and uniswap v3 arbitrum in the deployment.json though

good catch! I believe those are because of the changes in the master branch after I rebased this branch. They are fixed after another rebase.

> * Do we need to re-QA check the deployments that got fixed?

It'll be helpful if PS can take a look of those networks with fix.

> * Also, in Boba (WAGMI v3 Option), the `mintSupply` is negative. Shouldn't it be only >= 0. I don't think it makes sense to say that the minted supply on the chain is < 0

that's right. taking a look into this.

**Comment by dmelotik** [comment]
I asked @pietydespises if he could take a second look at aurora, celo, harmony, moonbeam, and boba

File: subgraphs/cbridge/README.md
Comment (dmelotik) [inline subgraphs/cbridge/README.md:1]: can we make a note that this covers cBridge v1/v2 contracts?

Comment (tnkrxyz) [reply subgraphs/cbridge/README.md:1]: My commit comment refers to [OriginalTokenVault](https://github.com/celer-network/sgn-v2-contracts/blob/main/contracts/pegged-bridge/OriginalTokenVault.sol), i.e. v1, and [OriginalTokenVaultV2](https://github.com/celer-network/sgn-v2-contracts/blob/main/contracts/pegged-bridge/OriginalTokenVaultV2.sol).  We included both in our current subgraph. I was not talking aobut cbridge version; 

Although they do have a sgn first version and corresponding contract [here](https://github.com/celer-network/sgn-contract/blob/master/contracts/SGN.sol), but I don't think itwas a functioning bridge yet.

**Comment by tnkrxyz** [comment]
> > * Also, in Boba (WAGMI v3 Option), the `mintSupply` is negative. Shouldn't it be only >= 0. I don't think it makes sense to say that the minted supply on the chain is < 0
> 
> that's right. taking a look into this.

OK, this was due to my mistake in the poolBridgeType argument of transfers on the originalTokenVault v1/v2. They are LOCK_RELEASE transfers, but I mistkenly labeled them BURN_MINT. This is now fix in a7c3483, and the new boba deployment shows [the mintSupply is now 0](https://subgraphs.messari.io/subgraph?endpoint=https%3A%2F%2Fapi.thegraph.com%2Fsubgraphs%2Fname%2Ftnkrxyz%2Fcbridge-boba&tab=pool&poolId=0x4c882ec256823ee773b25b414d36f92ef58a7c0cc6158b1989f89977bcc3150fc1f2eb2260f6cabe&view=poolHourlySnapshots-mintSupply).

@pietydespises , this change will affect the TVL, inputTokenBalance, and mintSupply. I am grateful if you can help take  quick look of these fields in all subgraphs. Sorry for not catching this sooner.

File: subgraphs/cbridge/README.md
Comment (dmelotik) [reply subgraphs/cbridge/README.md:1]: Gotcha, thanks for the clarity. I was thinking we were looking at 2 cBridge versions

**Comment by dmelotik** [comment]
@tnkrxyz @pietydespises for the revenue we can make a note about that in the README for future reference

**Comment by tnkrxyz** [comment]
@pietydespises , thx! I'm taking a look of the issues you found.

**Comment by tnkrxyz** [comment]
> Hey @tnkrxyz sorry for the delay here. Basically this looks great, with some small comments below:
 
> Boba metrics look wonky, showing zero TVL, but this might be explained by very low use. DeFiLlama showing ~$50k of TVL. But again they calculate it different than us, so might be OK.

This is because we don't have price oracle on boba network and all tokens have an effective priceUSD of 0. Even though the inputTokenBalance > 0, the totalValueLockedUSD will always be $0: https://api.thegraph.com/subgraphs/name/tnkrxyz/cbridge-boba/graphql?query=query+MyQuery+%7B%0A++pools+%7B%0A++++inputTokenBalance%0A++++inputToken+%7B%0A++++++name%0A++++++lastPriceUSD%0A++++%7D%0A++++totalValueLockedUSD%0A++++type%0A++%7D%0A%7D
 
> Aurora looks good
> 
> Celo metrics look wonky, showing zero TVL, but this might be explained by very low use. DeFiLlama showing only $100 worth of TVL so probably accurate.

Celo is a bit different. cbridge deployed a liquidity-based pool [there](https://celoscan.io/address/0xBB7684Cc5408F4DD0921E5c2Cadd547b8f1AD573#events), but there was never any LiquidityAdded events (filtering events by the signature of "LiquidityAdded( uint64 seqnum, address provider, address token, uint256 amount)" or 0xd5d28426c3248963b1719df49aa4c665120372e02c8249bbea03d019c39ce764 returns no results).

cbridge didn't deploy OriginalTokenVault (or v2) contract on celo (https://cbridge-docs.celer.network/reference/contract-addresses), this is the only other bridge pool type that has TVL. So I don't see how it has non-zero TVL.
 
> Harmony there's a sharp dropoff in transaction count in late June 2022. It looks like Harmony was probably taken off cBridge as an option (the option to bridge isn't there anymore when I check the app today), but then why is there still TVL showing up after those dates? And also still a very small amount of transactions occuring. Worth taking a look at I think.

The contract is still active, at least as of April 4th (https://explorer.harmony.one/address/0x78a21C1D3ED53A82d4247b9Ee5bF001f4620Ceec?activeTab=0), the last withdraw call was on March 11th. Unfortunately, the harmony explorer isn't providing an Events tab like etherscan, so I cannot search and find out the frequency of events or summarize them to validate,  but I am not surprise if there is non-zero TVL on the chain, even if the cbridge took it off their front end. 
 
> Moonbeam looks good.
> 
> Just asking again @tnkrxyz there's no way to make revenue look more smooth is there? I remember your explanation as to why it looks like that, but figured I asked again because we're going to need to add some sort of explanation in the product otherwise.

That is correct. For cbridge, we have no way to get the revenue number as it accrue on chain (because it is only available on the SGN network API), we can only track revenue as it is withdrawn [when the refId = `0x01`](https://github.com/celer-network/sgn-v2-contracts/blob/61159ed26e45e23731e4ed883b1d83be987d6c1a/contracts/pegged-bridge/OriginalTokenVault.sol#L49-L52). I added a sentence to README explaining the reason.

**Comment by tnkrxyz** [comment]
> @tnkrxyz @pietydespises for the revenue we can make a note about that in the README for future reference

added.

**Pull request merged**